Adding singularity command group with recipe|build

vsoch opened the pull request:
# Singularity Command

This is an early effort at a Singularity command, meaning that we create a
Singularity recipe (and can then optionall build a container) using it. It was fairly easy to put together with great example from https://github.com/spack/spack/pull/7204. This implementation is much simpler - we don't start any mirrors / volumes, we just scoot out a recipe and optionally build.

## The Command Group

Here is what the command group looks like. The user has the option to select
"recipe" or "build" subcommands. This extra grouping isn't essential, but provided
(as a design decision) so the final typed command is intuitive (reads like
a sentence).

```bash
$ bin/spack singularity -h
usage: spack singularity [-h] build|recipe ...

use singularity to build containers with spack packages

positional arguments:
  build|recipe
    recipe      generate a recipe
    build       generate a recipe and build it (required sudo)

optional arguments:
  -h, --help    show this help message and exit

```

For example, it's intuitive to ask for a recipe. 
Let's give a go to generate a recipe with a debian base, and install curl:

```bash
$ spack singularity recipe --from ubuntu:16.04 curl
==> Installing packages curl
==> Attempting to infer distro from base image...
==> Found distro ubuntu
==> Working directory created in /tmp/tmpfvp7wlfq
==> Writing recipe to /tmp/tmpfvp7wlfq/Singularity
==> sudo singularity build container.sif /tmp/tmpfvp7wlfq/Singularity
```

Here is the recipe generated from a template:

```
Bootstrap: docker
From: ubuntu:18.04

%environment
SPACK_ROOT=/home/spack
FORCE_UNSAFE_CONFIGURE=1
PATH="${PATH}:${SPACK_ROOT}/bin"
export SPACK_ROOT FORCE_UNSAFE_CONFIGURE PATH

%help
This is a spack container with packages curl

%post
SPACK_ROOT=/home/spack
FORCE_UNSAFE_CONFIGURE=1
PATH="${PATH}:${SPACK_ROOT}/bin"
export SPACK_ROOT FORCE_UNSAFE_CONFIGURE PATH

# Install basic system dependencies
apt-get update && \
    apt-get install -y git gcc g++ gfortran make bzip2 && \
    apt-get install -y patch file curl python gnupg2 xz-utils && \
    rm -rf /var/lib/apt/lists/*
    

# Install spack + set the mirror
mkdir -p $SPACK_ROOT
curl -s -L https://api.github.com/repos/spack/spack/tarball/develop \
    | tar xzC $SPACK_ROOT --strip 1

/home/spack/bin/spack install curl

%runscript
exec /home/spack/bin/spack "$@"
```

The final line is the command (instruction) to the user to build the container.
Let's do that.

```bash
$ sudo singularity build container.sif /tmp/tmpfvp7wlfq/Singularity
```

Then my container runs, exposing the spack entrypoint (and I can find curl).

```bash
$ singularity run container.sif 
usage: spack [-hkV] [--color {always,never,auto}] COMMAND ...

A flexible package manager that supports multiple versions,
configurations, platforms, and compilers.

These are common spack commands:

query packages:
  list                  list and search available packages
  info                  get detailed information on a particular package
  find                  list and search installed packages

build packages:
  install               build and install packages
  uninstall             remove installed packages
  spec                  show what would be installed, given a spec

environments:
  env                   manage virtual environments
  view                  project packages to a compact naming scheme on the filesystem.

modules:
  load                  add package to environment using `module load`
  module                manipulate module files
  unload                remove package from environment using `module unload`

create packages:
  create                create a new package file
  edit                  open package files in $EDITOR

system:
  arch                  print architecture information about this machine
  compilers             list available compilers

optional arguments:
  -h, --help            show this help message and exit
  -k, --insecure        do not check ssl certificates when downloading
  -V, --version         show version number and exit
  --color {always,never,auto}
                        when to colorize output (default: auto)

more help:
  spack help --all       list all commands and options
  spack help <command>   help on a specific command
  spack help --spec      help on the spec syntax
  spack docs             open http://spack.rtfd.io/ in a browser
vanessa@vanessa-ThinkPad-T460s:~/Documents/Dropbox/Code/Python/spack$ singularity run container.sif find curl
==> 1 installed package
-- linux-ubuntu18.04-x86_64 / gcc@7.4.0 -------------------------
curl@7.63.0
```

That's basically it - it's the simplest implementation I could think of, meaning
I don't bring in any special libraries (Singularity Python, HPC Container Maker, etc.)
The user can switch recipe --> build (and use sudo) to do the same process, but
building the container at the end.

```
$ sudo bin/spack singularity build --from ubuntu:16.04 curl
```

For the interested reader, here is an asciinema that shows the recipe generation
and then build (to confirm that the example at least works!)

[asciinema](https://asciinema.org/a/244183?speed=3)


## Discussion

### Should containers be kept with spack?

Given that we have a build command, is there some avenue for containers to be "a thing"
that spack installs? For example, instead of having `opt/spack/linux-ubuntu16.04-x86_64/`
We might dump a container into `opt/spack/containers/singularity` and then name based on
the packages installed (the specs.)


### Entrypoint examples?
Having spack as the entrypoint makes sense for discovery of what is installed. But 
what is the way I would specify running curl (or any installed package) via the spack
entrypoint? We should have these examples provided (by default) in the container
help.

### Alpine / busybox base

<s>We need to have this as an option - I'll just start working on it (need to test
the dependencies for spack) and then give it a go (not sure if spack supports these
bases, but I'll find out!)</s>

alpine is added. I'm not sure we need busybox, but others can comment.

### Custom environment / labels / files

What labels would be useful to add (by default?) Is it a valid use case to want
to add some other set of environment variables, labels, files, or commands?

### Reproducibility

The "reproducibility raccoon" in me wants to (given that the user builds the container
for storage with spack) also keep the recipe, especially if we open up avenues to
add further customization other than installing packages.

### Customize entrypoint?

Any potential use case of a user wanting the entrypoint to not be spack?
I would err on not providing this because we lose information / metadata about
the container that spack can provide. But it could be wanted.

### Don't generate recipe?
Arguably, if we don't need to build, the recipe can be printed to screen.
I am assuming that the average user generating a recipe wants to keep it,
so I wrote to file. However it would be one extra step to pipe into a recipe file,
in case that it's wanted. I would say this is a reasonable start unless someone
has huge issue.

### Check if Singularity not installed

Does spack do this when I create the executable (or do I need to?)

Signed-off-by: Vanessa Sochat <vsochat@stanford.edu>

vsoch commented:
Just had a nutty idea - technically someone could use a list of the packages and build a container for each one. And even just provide for download from some https:// location directly to singularity! That would be pretty neat :)

vsoch commented:
Just added a base recipe for alpine, and also the template was embarrassingly in the wrong folder (I had a merge conflict and put together a quick new repo to make the PR, and should have double checked my move operations).

vsoch commented:
Where are tests written for the commands?

vsoch commented:
Looks like we start in this qa folder:

```
$ ls share/spack/qa/
bashcov  configuration  run-build-tests  run-docker-tests  run-doc-tests  run-flake8-tests  run-unit-tests  setup.sh
```
And then it uses spack itself to run test, the "test" command:

```bash
${coverage_run} bin/spack test --verbose "$extra_args"
```

Ahh I just found a hint in .coveragerc
```
$ cat .coveragerc 
# -*- conf -*-
# .coveragerc to control coverage.py
[run]
parallel = True
concurrency = multiprocessing
branch = True
source = lib
omit =
     lib/spack/spack/test/*
     lib/spack/docs/*
     lib/spack/external/*
```
and I'd guess the test files for commands are here?

```bash
$ ls lib/spack/spack/test/cmd/
activate.py    cd.py        dependencies.py  gpg.py      license.py           providers.py     test_compiler_cmd.py
arch.py        clean.py     dependents.py    graph.py    list.py              python.py        uninstall.py
blame.py       commands.py  env.py           help.py     mirror.py            release_jobs.py  url.py
buildcache.py  config.py    find.py          info.py     module.py            resource.py      versions.py
build_env.py   debug.py     flake8.py        install.py  print_shell_vars.py  spec.py          view.py
```

vsoch commented:
How do I run a single test?

vsoch commented:
hey @hartzell can I recruit you to help me work on this? :) You're super great to work with (come baaaack)!

alalazo commented:
> Where are tests written for the commands?

Tests for the built-in commands are in `lib/spack/spack/test/cmd`, but I see that you already figured that out :blush: 

To run tests you can use:
```console
$ spack test
```
which is a wrapper around `pytest.main()`. To get all the `pytest` options you can do:
```console
$ spack test -H
usage: spack test [options] [file_or_dir] [file_or_dir] [...]

positional arguments:
  file_or_dir

general:
  -k EXPRESSION         only run tests which match the given substring
[...]
to see available markers type: pytest --markers
to see available fixtures type: pytest --fixtures
(shown according to specified file_or_dir or current dir if not specified)
```
and to run a single test one method is:
```
$ spack test -k <test_name>
```
Other useful combinations when testing:
```
# Runs the tests that failed before first
$ spack test --ff

# Runs only the tests that failed before [and allow at most 'num' failures]
$ spack test --lf [--maxfail=num]
```

vsoch commented:
Thank you! I'll mess around for a bit, and keep pressing keys until something is working (and then come back!) :)

vsoch commented:
okay, we have a start! I was able to get the test file running, specifically, the file `lib/spack/spack/test/cmd/singularity.py` by doing:

```bash
$ bin/spack test -k singularity
```
and then a bunch of spaghetti hit the screen (I hadn't written any actual tests yet!) I started, but then I realized to my own horror that I hadn't written docs yet. So I just added the start to those (I was going to create a "containers" group anticipating the other PR with HPC container maker, but decided to keep it simple and add the `spack singularity` command group to `basic_usage`. Is there anywhere else that the library goes through commands, or is this good to start? Whenever I have readthedocs that are a bit robust (akin to this one) I would just do a search anyway.

I'm going for a quick run but I'll be back in a bit to get back to those tests!

vsoch commented:
Feel free to cancel the first run of testing (since I pushed a trivial change) and maybe if Travis has the equivalent of "cancel redundant builds" we could turn that on?

truatpasteurdotfr commented:
Naive question: wouldn't it be easier to use a "fixed" development singularity container and bind the /slack tree with a discardable $HOME ?

vsoch commented:
No I don’t think so, because the original time to get the base is trivial, and then we would have to think about providing bases (which one/s?) and also versions. This also assumes the user wants to build, period, which may not be the case. -1 from me.

vsoch commented:
@truatpasteurdotfr if the user is building with Singularity on their machine, it's still going to use cached layers. The reason we can't do some caching within the container with spack is well articulated here https://github.com/spack/spack/pull/10952#issuecomment-490092643. But thanks for your suggestion!

vsoch commented:
Okay I've just pushed some updates!

 - I made a first (tiny) effort at tests. I realize that we likely can't use sudo during testing (or can we?) so the most I'm able to do is check that spack errors out without it. Given that we can't test build, this means we can't test the actual build of the recipes. But we can check the generation for different bases. This might not be a terrible thing, because it takes a long time. 
 - I realized that the user will likely want control of the output directory, so I added the argument working_dir to customize that. 
 - I added more to the docs to review all of the options / args for the singularity command. 
 - I checked flake8 this time first :tada: :) 
 - I didn't check building the docs, it pooped out with a title not being long enough (doh!)
 
I'm not familiar with doing / using mocks, so for additional tests that are desired I will need help (examples to get me started). Let me know how things look! Holy cow, there are a crapton of tests! I'll check back in a bit to see if there are any other tweaks to make.

vsoch commented:
Are the docs cached? I fixed the Singularity header that was too short, but it pooped out again.

vsoch commented:
Ah, hold the phone! I was looking at the wrong tab. It looks like I forgot to add a blank line after a code block declaration in the docs -

```
Warning, treated as error:
/home/travis/build/spack/spack/lib/spack/docs/basic_usage.rst:1317:Error in "code-block" directive:
maximum 1 argument(s) allowed, 23 supplied.
```

just pushed a fix!

vsoch commented:
okay the docs issues are fixed, and we are in the state of "the build timed out for no apparent reason because Travis didn't eat his wheaties this morning." I'll await further instruction.

vsoch commented:
ping! I'm not sure how to address the (one tiny failing bullet) about the coverage. Let me know, and happy Friday!

vsoch commented:
Is anyone going to review this?

scheibelp commented:
> Is anyone going to review this?

@alalazo self-assigned this so I assumed he would look at it eventually

> I'm not sure how to address the (one tiny failing bullet) about the coverage.

Your coverage is pretty good so that's not an issue IMO. That being said as of https://github.com/spack/spack/pull/11411 I was expecting to be able to see the actual line coverage, which would have let me suggest what else to test. But as of now I still cannot access them. You can run the unit tests locally and generate coverage stats from them (although like I said the percentage looks good so that isn't necessary).

Also, I asked in https://github.com/spack/spack/pull/10952#issuecomment-492466323 how this differs from #10952: the approaches are definitely different but I was also curious about features.

vsoch commented:
This implementation does not depend on a third party (nvidia owned) package. It does seem trivial to pip install something, but when I develop I tend to try and minimize extra dependencies (keeping them to zero, if possible) to put the smallest additional burden on using the software. Thinking of the future, every additional library that is added is making a strong assertion that the library will be maintained, reliable, have no license change, and something along the lines of best practice for performing its function. For this reason, I think it's logical that there is no additional dependency to generate a Singularity recipe, and the only dependency to build a Singularity recipe is... Singularity.  In terms of thinking about this from the developer point of view, if I'm a spack developer in the future and there is issue with the code, it's double the work to have to triage the dependency repository first, wait for a release, and then open a pull request here. Why would I do that unless absolutely necessary?

In terms of features, I decided to keep it minimal again so that the installation of software would be done by Spack - not by apt-get, etc. in the container. That said, if spack itself requires other bases on the host for different kinds of install, those should be added as options (and if you could point me to those, it would be greatly appreciated). The other feature that I see as reasonable would be taking as input any number of the various spack input files (I'm not hugely familiar again, but I've seen these packages.yml around and think that might be one!) However, I also advocate for not over-engineering and adding feature bloat before it's wanted. It's easy enough to add a skeleton, and add additional features when they are requested by the user community.

These are the reasons I have chosen to take this approach. Let me know if you want to discuss any needed features mentioned above, or have other questions, and thanks for the update!

alalazo commented:
@vsoch Hey, apologies for being silent - have been busy preparing [this](https://alalazo.github.io/spack-cscs2019/) and associated container for a demo. Will try to review this PR by tomorrow, end of the working day.

vsoch commented:
Hey no worries - good luck on your presentation!

vsoch commented:
And just a quick note - I don't see the situation as being a "this or that" between this pull request and #10952. They are different. I might suggest, however, that if hpccm is useful to some cohort for spack, and the extra burden of taking on another dependency is wanted, it isn't added as a "containers" command but more directly added as a `spack hpccm` command, or something that directly r eferences the tool tied into. This means that someone could also then add a `spack docker` or (insert technology of choice) command. I would leave the "containers" command for (down the line) when there is better tooling for some consistent generation of a container that, maybe even the software / spec isn't created yet! Just 0.02. :)

alalazo commented:
A few random comments after reading this PR, mostly to trigger a discussion.

#### Support for container in Spack

There are now three PRs that add support for containers in some form: #11367, #10952 and #7204. We probably need to sync and rationalize the efforts, so that when we use those commands the interface and functionality will look consistent. What I like of this PR is it's simplicity - `build` and `recipe` should probably be also available in the same form in #7204 (or a reworking of it).

One thing that I wonder is if container functionality should stay in `spack` core or should be supplied as an extension command (see #11209). This would basically mean moving this piece of code (and possibly that of the other 2 PRs) into another repository within the Spack organization. It would be a first and I see both pros and cons to this decision, so probably @tgamblin is the right person to have a say on this idea.

#### Entrypoint examples + Custom environments

> But what is the way I would specify running curl (or any installed package) via the spack entrypoint?

We should probably try to integrate Spack environments with containers and either:

1. Set views in an appropriate way to make applications available
2. Activate that environment and leave Spack in the final image

Ideally in the long run we should be able to install the software we need with Spack and get rid of all build dependencies. In the case we set views we could also get rid of Spack itself using a multi-stage build. This should reduce the footprint of the image in case it's needed. Maybe strip binaries too.

> What labels would be useful to add (by default?) Is it a valid use case to want to add some other set of environment variables, labels, files, or commands?

The best solution that comes to my mind for custom environment variables is to add them to Spack environments (the functionality currently is not there) and use that information in some way when we build a container. 

Using Spack environments as a driver for the installations should also help in some sense with reproducibility, as ideally all the information should be in the manifest file.

#### Customize entry point

What is the use-case we are targeting? If it is to satisfy a user that wants to install a given set of applications or libraries in an image probably the best entry point is a shell with the applications in the PATH?

#### Check if Singularity not installed

To do this you can use:
```python
import spack.util.executable

singularity = spack.util.executable.which('singularity')

if singularity is None:
   # Handle not found
   pass
```

vsoch commented:
One thing that I wonder is if container functionality should stay in spack core or should be supplied as an extension command (see #11209). This would basically mean moving this piece of code (and possibly that of the other 2 PRs) into another repository within the Spack organization. It would be a first and I see both pros and cons to this decision, so probably @tgamblin is the right person to have a say on this idea.

Entrypoint examples + Custom environments
But what is the way I would specify running curl (or any installed package) via the spack entrypoint?

We should probably try to integrate Spack environments with containers and either:

## Extensions

This is a really cool idea! I agree that @tgamblin can have the ultimate weigh in, but I'll give some feedback. I would logically divide something that would be an extension based on dependencies:

 - if it requires additional install of software, make an extension
 - it it doesn't, provide with spack

This Singularity group would likely be borderline, because it can be used without extra dependencies for a recipe, but you of course would need Singularity to build. The other thing we should consider is ease of use for the user. If I download spack, if it's included natively I'm ready to go. If not, I'd look for some singularity command, not find it, and either give up  or need to know / search for how to install extensions. This makes me think that another variable to add to the "should it be an extension" or not is how widely used the command would be. I defer to @tgamblin! T

## Views

I'm not totally familiar with views - what do you mean "set views?" It sounds like this is some data structure that we could go back to and remove build dependecies (akin to multi-stage build?) If so, I completely agree! I think this should be handled by spack too - it shouldn't be that the container / container generation software needs to figure out what to add and remove, it should be possible to derive via spack, and then spit out the container generation command to copy the final folders/files that are needed. Actually, thinking of it now, that's a super cool (and relatively straight forward) thing to do!
Are views currently available? I could give a go at implementation.

## Docker 

I know this is a Singularity command, but I'll point out that the simple approach I took could also be easily done with Dockerfiles. It's just using a different recipe template. I don't see Docker as being a huge need for HPC so I didn't work on it.

## Environment / Labels

> The best solution that comes to my mind for custom environment variables is to add them to Spack environments (the functionality currently is not there) and use that information in some way when we build a container. Using Spack environments as a driver for the installations should also help in some sense with reproducibility, as ideally all the information should be in the manifest file.

So this isn't implemented yet? Is there a PR underway? I'm guessing this is some environment file that we could provide as a parameter? *clicks staples easy button* we can do that!

## Entrypoint

> What is the use-case we are targeting? If it is to satisfy a user that wants to install a given set of applications or libraries in an image probably the best entry point is a shell with the applications in the PATH?

Currently, the entrypoint is just spack. So when you interact with the container, you hit spack. It could be that the user wants the container to serve curl directly (or some other software inside). Let's say that the entrypoint is spack - how would the user run curl? Is there a "spack run" akin to [scif](https://sci-f.github.io/tutorial-really-quick-start#Run)? Something like:

```bash
$ spack run curl
```

## Check singularity installed

Let me add this asap!

The other question I was curious about is what input files a user might want to add to install from (e.g., packages.yml?) If that would be desired, could you point me to examples for how to do that?

Will push an update soon!

vsoch commented:
okey doke! Just updated with develop, and added a function to check if Singularity is installed, only needed in the case of build. Here I removed Singularity and ran build:

```bash
$ sudo spack singularity build --from ubuntu:16.04 curl
==> Installing packages curl
==> Attempting to infer distro from base image...
==> Found distro ubuntu
==> Working directory created in /tmp/tmpR448yH
==> Writing recipe to /tmp/tmpR448yH/Singularity
==> Error: singularity must be installed for build. Try recipe.
```

samcmill commented:
> ## Docker
> I know this is a Singularity command, but I'll point out that the simple approach I took could also be easily done with Dockerfiles. It's just using a different recipe template. I don't see Docker as being a huge need for HPC so I didn't work on it.

As a container runtime, Docker is not relevant for HPC.  But as a container image builder, image format (OCI), and image repository, Docker is very relevant to HPC.  Shifter and Charliecloud both rely on Docker (OCI) images and repositories, for instance.

vsoch commented:
Good call! I’d be happy to add a template/command to generate recipes for Docker, after we nail down how we want to tweak this one.

alalazo commented:
> This Singularity group would likely be borderline, [...] how widely used the command would be. I defer to @tgamblin! T

I think we are really on the same page here. The only thing I'd like to add is that we can mitigate the downside of having to know how to install an extension by providing a core command along the lines of `go get`. Say we have an extension `spack-containers` that provides all the subcommands related to supported containers, and that this extension is officially maintained in the Spack organization. I imagine that a user could install that extension via cli by doing something like:
```console
$ spack get containers
[...]
```
i.e. in a single command. That's for sure still one command more than none, but much better than downloading the extension and editing a YAML file manually.

vsoch commented:
Yeah that's pretty cool :) 

I'd definitely be down to make some extensions - they would be separate repos under the spack org? Say the word, and I'm on it! (and I'll likely need to be a member to create a repo).

alalazo commented:
> I'm not totally familiar with views - what do you mean "set views?" [...] Actually, thinking of it now, that's a super cool (and relatively straight forward) thing to do!
Are views currently available? I could give a go at implementation

As you might have noticed I am answering to your comment in bits and pieces :slightly_smiling_face:

This bits starts with a small detour.

Last November we added to Spack a major feature, Spack environments. Thery're basically a way to declare in a *manifest* file the user's abstract requirements (i.e. the specs he wants installed) and let Spack manage the creation of a *lock* file when you decide to install on a given platform - [docs here](https://spack.readthedocs.io/en/latest/tutorial_environments.html).  The feature still needs some polishing, see #11372 and #11057, but is already in a state where it can be useful. By default Spack environments set "views", which basically means that everything installed in the environment will be symlinked in a single prefix that is laid out in the usual way you expect in a distro. 

I'll use an example from a recently given talk to be to be clearer. Here is the view of an environment where I installed all the specs that are needed to start developing a package called `hpx` right away:
```console
cscs-smc-2019 [~/hpx-development/deps] $ tree -d -L 1 .spack-env/view/
.spack-env/view/
|-- bin
|-- doc
|-- etc
|-- include
|-- lib
|-- lib64
|-- man
|-- sbin
`-- share

9 directories

$ tree  -L 1 .spack-env/view/bin
.spack-env/view/bin
|-- 2to3
|-- bsdcat -> /home/spack/spack/opt/spack/linux-ubuntu16.04-x86_64/gcc-8.3.0/libarchive-3.3.2-2ys2japbdbwad55onhotwzzmr5djwgut/bin/bsdcat
|-- bsdcpio -> /home/spack/spack/opt/spack/linux-ubuntu16.04-x86_64/gcc-8.3.0/libarchive-3.3.2-2ys2japbdbwad55onhotwzzmr5djwgut/bin/bsdcpio
|-- bsdtar -> /home/spack/spack/opt/spack/linux-ubuntu16.04-x86_64/gcc-8.3.0/libarchive-3.3.2-2ys2japbdbwad55onhotwzzmr5djwgut/bin/bsdtar
[...]
|-- unxz -> /home/spack/spack/opt/spack/linux-ubuntu16.04-x86_64/gcc-8.3.0/xz-5.2.4-b3wc6au4psflllc3w5yax4zw5ypn3dhu/bin/unxz
|-- valgrind -> /home/spack/spack/opt/spack/linux-ubuntu16.04-x86_64/gcc-8.3.0/valgrind-3.14.0-evumvc5rjg5kxf3mys2rgm6iw4eddjhb/bin/valgrind
|-- valgrind-di-server -> /home/spack/spack/opt/spack/linux-ubuntu16.04-x86_64/gcc-8.3.0/valgrind-3.14.0-evumvc5rjg5kxf3mys2rgm6iw4eddjhb/bin/valgrind-di-server
[...]
```
As you can see we basically projected into the view a subset of Spack's store. Also, when a user *activates* an environment the `bin` folder gets added to PATH, pretty much like in Python's virtual environments.

Now, coming back to your questions (or my understanding of them). What I was suggesting is that, in addition to taking specs directly, any command that builds container images should be able to receive as an alternative input a *manifest* file. In this way the build command will be just `spack install`, no matter how complex the list of things you want to build in the image. Then, once an image is built from an environment we could either:

1. Configure the view to be created in a convenient place (e.g. /opt/\<any-name-we-like\>) and add the `bin` subfolder to PATH, or
2. Activate the environment, so that the step above will be managed by Spack instead of in the recipe to build the container image

This of course implies that the entry point is not Spack, but a shell. Does that make sense to you? The idea to remove build dependencies, or strip binaries would be a later optimization - but I think it's worth start considering it.

> The other question I was curious about is what input files a user might want to add to install from (e.g., packages.yml?) If that would be desired, could you point me to examples for how to do that?

I think it would be only `spack.yaml`, as it contains all the available configurations.

vsoch commented:
I think I get it! Let me give a go at activating a spack environment, then installing, and then having the entrypoint be a regular shell (and it should find curl, etc.) Be back in a bit!

vsoch commented:
okay this is really cool! So I added another argument (if the user has preference for a name) to install to a spack environment. The default is `spackenv` (super creative, I know :P). Then when we do the build, before install, we activate it:

```bash
==> Updating view at /home/spack/var/spack/environments/spackenv/.spack-env/view
==> Created environment 'spackenv' in /home/spack/var/spack/environments/spackenv
+ /home/spack/bin/spack env activate spackenv
==> This command works best with Spack's shell support
  
  To initialize spack's shell commands:
  
      # for bash and zsh
      . /home/spack/share/spack/setup-env.sh
  
      # for csh and tcsh
      setenv SPACK_ROOT /home/spack
      source /home/spack/share/spack/setup-env.csh
  
  Or, if you want to use `spack env activate` without initializing
  shell support, you can run one of these:
  
      eval `spack env activate --sh spackenv`   # for bash/sh
      eval `spack env activate --csh spackenv`  # for csh/tcsh
+ /home/spack/bin/spack install perl
==> Installing pkgconf
```
And then the package installs as usual. I can shell into the container, and try different ways to activate.

```
%runscript
/home/spack/bin/spack env activate spackenv
```

But it doesn't seem to find the perl install:

```
Singularity container.sif:/tmp/tmprvEL4l> spack find
==> In environment spackenv
==> No root specs

==> 0 installed packages
```
So I did a little more testing, let's try again.

```bash
> spack env activate spackenv
==> This command works best with Spack's shell support
  
  To initialize spack's shell commands:
  
      # for bash and zsh
      . /home/spack/share/spack/setup-env.sh
  
      # for csh and tcsh
      setenv SPACK_ROOT /home/spack
      source /home/spack/share/spack/setup-env.csh
  
  Or, if you want to use `spack env activate` without initializing
  shell support, you can run one of these:
  
      eval `spack env activate --sh spackenv`   # for bash/sh
      eval `spack env activate --csh spackenv`  # for csh/tcsh
```
okay I can try that:

```bash
. /home/spack/share/spack/setup-env.sh
```
I can find the software I installed:

```bash
Singularity container.sif:/tmp/tmprvEL4l> spack find
==> 5 installed packages
-- linux-alpine3-x86_64 / gcc@6.4.0 -----------------------------
gdbm@1.18.1  ncurses@6.1  perl@5.26.2  pkgconf@1.6.0  readline@7.0
```

but I could do that before. There isn't anything added to the PATH (and it's not changed)
```
> which perl
```
I took a look at the spack.yaml, it looks like there isn't anything installed here, so activation during build and then install didn't seem to work? Did I do that wrong?

```bash
> cat /home/spack/var/spack/environments/spackenv/spack.yaml 
# This is a Spack Environment file.
#
# It describes a set of packages to be installed, along with
# configuration settings.
spack:
  # add package specs to the `specs` list
  specs: []
  mirrors: {}
  modules:
    enable: []
  repos: []
  packages: {}
  config: {}
  upstreams: {}
```
I think this is likely the issue. I was following instructions in the doc that you linked, but it doesn't seem to install. Could you provide the correct steps to (headless) create the environment, install to it, and then activate, and show perl on the path? Per your example with the view folder, I'm also not sure where that is. It could be an issue that it gets generated in $HOME, but then a regular user can't see it (in root) and if the container is run as root it gets bound over. Actually, I can test this!

```
$ sudo singularity shell --containall --no-home container.sif 
```
and I found a spack thing!

```bash
Singularity container.sif:~> ls /root/.spack/
cache/ linux/ 
```
but I'm not sure if there are environments in there. Any help appreciated!

vsoch commented:
I'm going to try building in Docker (with write) and using the "spack add" command after activating the environment.

vsoch commented:
Okay, trying the add is helpful. I'm able to add perl to it, and then get a visual for where the view directory is:

```
bash-4.4# spack env activate spackenv
bash-4.4# which perl
bash-4.4# spack add perl
==> Adding perl to environment spackenv
==> Updating view at /home/spack/var/spack/environments/spackenv/.spack-env/view
```
It's still not added to any PATHS
```
bash-4.4# which perl
bash-4.4# 
bash-4.4# echo $PATH
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/spack/bin
```
but now I can see it added to the environment file:

```
bash-4.4# cat /home/spack/var/spack/environments/spackenv/spack.yaml 
# This is a Spack Environment file.
#
# It describes a set of packages to be installed, along with
# configuration settings.
spack:
  # add package specs to the `specs` list
  specs: [perl]
  mirrors: {}
  modules:
    enable: []
  repos: []
  packages: {}
  config: {}
  upstreams: {}
```
Now it appears in root specs:

```
bash-4.4# spack find
==> In environment spackenv
==> Root specs
perl

==> 0 installed packages
```
But not installed

```
bash-4.4# spack find perl
==> No package matches the query: perl
```
Let's try as the docs suggest, do another "spack install"

```
bash-4.4# spack install
==> Concretizing perl
[+]  zd4pl2i  perl@5.26.2%gcc@6.4.0+cpanm patches=0eac10ed90aeb0459ad8851f88081d439a4e41978e586ec743069e8b059370ac +shared+threads arch=linux-alpine3-x86_64
[+]  k26okw7      ^gdbm@1.18.1%gcc@6.4.0 arch=linux-alpine3-x86_64
[+]  3p75lon          ^readline@7.0%gcc@6.4.0 arch=linux-alpine3-x86_64
[+]  k5ix7ys              ^ncurses@6.1%gcc@6.4.0~symlinks~termlib arch=linux-alpine3-x86_64
[+]  zdptkwp                  ^pkgconf@1.6.0%gcc@6.4.0 arch=linux-alpine3-x86_64
==> Updating view at /home/spack/var/spack/environments/spackenv/.spack-env/view
==> Installing environment spackenv
==> perl is already installed in /home/spack/opt/spack/linux-alpine3-x86_64/gcc-6.4.0/perl-5.26.2-zd4pl2i7uhm6kwvjwp7ijonww6ggcprm
==> Updating view at /home/spack/var/spack/environments/spackenv/.spack-env/view
```
Ahh now the view is updated! And we have a bin, etc.

```bash
bash-4.4# ls /home/spack/var/spack/environments/spackenv/.spack-env/view/
bin      include  lib      share
bash-4.4# ls /home/spack/var/spack/environments/spackenv/.spack-env/view/bin/
captoinfo         encguess          infocmp           ncursesw6-config  perlthanks        pod2usage         ptargrep          toe
clear             gdbm_dump         infotocap         perl              piconv            podchecker        reset             tput
corelist          gdbm_load         instmodsh         perl5.26.2        pl2pm             podselect         shasum            tset
cpan              gdbmtool          json_pp           perlbug           pod2html          prove             splain            xsubpp
cpanm             h2ph              libnetcfg         perldoc           pod2man           ptar              tabs              zipdetails
enc2xs            h2xs              ncurses6-config   perlivp           pod2text          ptardiff          tic
```
Still not added to the path (automatically) - and I tried activating again in multiple ways - BUT I can easily just add this to the path.

alalazo commented:
Just to clarify, in case it's needed. The `spack.yaml` file can be created outside of the container. What I was thinking is to copy it in the container during the build. If `spack.yaml` is in `pwd` then the environment is implicitly activated, and a single `spack install` should trigger all the necessary installations.

vsoch commented:
can you give me a spack.yml you have lying around to test? I actually just got the (internally created) environment working, so that the user can enter the container (via shell) and perl / whatever is in the views bin is on the path. I have the Singularity recipe working, and here is the Dockerfile I was using to test:

```
FROM vanessa/salad

# docker build -t spack/spack .
# docker run -it --entrypoint bash spack/spack

ENV SPACK_ROOT=/home/spack
ENV FORCE_UNSAFE_CONFIGURE=1
ENV PATH="${PATH}:${SPACK_ROOT}/bin"

# Install basic system dependencies
RUN apk update && \
    apk add --no-cache git gcc g++ gfortran make bzip2 && \
    apk add --no-cache patch file curl python gnupg xz && \
    apk add --no-cache curl bash openssh libtool linux-headers
    

# Install spack + set the mirror
RUN mkdir -p $SPACK_ROOT && \
    curl -s -L https://api.github.com/repos/spack/spack/tarball/develop \
        | tar xzC $SPACK_ROOT --strip 1

RUN spack env create spackenv && \
    spack install perl && \
    echo ". /home/spack/share/spack/setup-env.sh" >> $HOME/.bashrc && \
    spack -e spackenv add perl && \
    spack -e spackenv install

# Add to Paths
ENV LD_LIBRARY_PATH=/home/spack/var/spack/environments/spackenv/.spack-env/view/lib:$LD_LIBRARY_PATH
ENV PATH=/home/spack/var/spack/environments/spackenv/.spack-env/view/bin:$PATH
```
Is this functionality wanted? It's different from taking the spack.yml file, which I'd expected only an advanced user to have. If you have an example one, I can try testing it out to provide as the specs parameter.

Now that I've done it, the functionality of shelling inside and having the software on the path is much better than needing to know how to find it with spack:

```bash
$ docker run -it --entrypoint bash spack/spack
bash-4.4# which perl
/home/spack/var/spack/environments/spackenv/.spack-env/view/bin/perl
```
Yeah I really like that :)  I won't push any updates until I've tested with an example environment, and also confirmed this is what you had in mind.

vsoch commented:
Using the yaml file would look like this:

```
COPY spack.yaml /spack.yaml
RUN spack env create spackenv && \
     cp /spack.yaml /home/spack/var/spack/environments/spackenv/spack.yaml && \
     echo ". /home/spack/share/spack/setup-env.sh" >> $HOME/.bashrc && \
     spack -e spackenv install
```
Do you really think this is a likely use case? It seems hard to create the file to begin with, easier to just list specs as strings. Is there more that the environment file can offer, other than a list of packages?

vsoch commented:
And just confirmed (the first approach) works with the Singularity template:

```
$ singularity shell container.sif 
Singularity container.sif:/tmp/tmpfVjiCt> which perl
/home/spack/var/spack/environments/spackenv/.spack-env/view/bin/perl
```

vsoch commented:
okay, I think I have it working both ways. Here is without a spack.yaml (and list of packages)

```
$ sudo bin/spack singularity build --from vanessa/salad --distro alpine perl
...
$ singularity shell /tmp/tmpVDPNRY/container.sif
Singularity container.sif:~/Documents/Dropbox/Code/Python/spack> which perl
/home/spack/var/spack/environments/spackenv/.spack-env/view/bin/perl
```
and here it is with the spack.yaml:

```bash
$ cat spack.yaml 
# This is a Spack Environment file.
#
# It describes a set of packages to be installed, along with
# configuration settings.
spack:
  # add package specs to the `specs` list
  specs: [perl]
  mirrors: {}
  modules:
    enable: []
  repos: []
  packages: {}
  config: {}
  upstreams: {}
```
```bash
$ sudo bin/spack singularity build --from vanessa/salad --distro alpine spack.yaml 
...
$ singularity shell /tmp/tmpBWLt8u/container.sif
Singularity container.sif:~/Documents/Dropbox/Code/Python/spack> which perl
/home/spack/var/spack/environments/spackenv/.spack-env/view/bin/perl
```

vsoch commented:
Is this what you had in mind?

vsoch commented:
hey @alalazo I have all these changes locally - let me know if you want them pushed here, or to another branch for inspection.

alalazo commented:
> Is this what you had in mind?

Yes. The single `spack install` should work **if** the command is issued from the directory in which `spack.yaml` resides. In that case the environment is activated implicitly so there should be no need for other commands.

> Do you really think this is a likely use case? It seems hard to create the file to begin with, easier to just list specs as strings. Is there more that the environment file can offer, other than a list of packages?

I think they're both valid use cases. With `spack.yaml` you can ship an entire configuration (generation of module files, setup of compilers, etc.) in a single file - and that's very convenient. But I agree that if you need just a couple of specs and you are fine with the defaults for other things accepting specs from the cli is the way to go. 

> let me know if you want them pushed here, or to another branch for inspection.

I'd wait to see what @tgamblin and other maintainers think about what is being discussed here. In particular if they think `spack singularity` should be a new built-in command or if instead it should be distributed as an extension.

vsoch commented:
Alrightly, I'll hold off on further updates. If you want to test, I think I need some help because I couldn't get the clean install (without specifying the environment) to work. The branch with these changes I just pushed here - https://github.com/researchapps/spack/tree/add/cmd-singularity-spack-yaml

If you need me to open another PR just to see the changes more easily let me know!

vsoch commented:
hey @alalazo this is now over 3 months old... any updates?

alalazo commented:
@vsoch Apologies for being silent for so long. There have been no updates yet on this topic (I'm mostly focused on deadlines for #3206 and #11372 + other bugfixes). Anyhow - this PR is still on my radar and I'll surely get back to this in early September. Would that work for you?

vsoch commented:
Yes of course! 🦖
